### PR TITLE
[💳] Saving credit cards

### PR DESCRIPTION
**Pull request by eoji**
# What
Users can now save credit cards to their account 🆒 

# How
- Updated Stripe library.
- Updated strings.
- A little renaming 😅 
- Refactored payment method row because it was breaking XML previews in Studio and it was just easier to do it as a LinearLayout 👀 
- Created card using Stripe SDK.
- Added mutation for saving cards.
- Added placeholder Stripe keys.
- Added `MockStripe` and `CardFactory` classes for testing ✨ 
- I'm doing a lot of shady things because the Stripe `CardInputWidget` doesn't properly tell me when it has focus or properly determine if a card is valid (IMO~)

# see

image

### CODE REVIEW

**Comment by eoji** [comment]
[Add new card screen](https://trello.com/c/Xz9KVA3V/1022-add-new-card-screen)

File: app/src/main/java/com/kickstarter/viewmodels/NewCardFragmentViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        private val postalCode = PublishSubject.create<String>()
+        private val saveCardClicked = PublishSubject.create<Void>()
+
+        private val cardWidgetFocusDrawable = BehaviorSubject.create<Int>()
+        private val error = BehaviorSubject.create<String>()
+        private val progressBarIsVisible = BehaviorSubject.create<Boolean>()
+        private val saveButtonIsEnabled = BehaviorSubject.create<Boolean>()
+        private val success = BehaviorSubject.create<Void>()
+
+        val inputs: NewCardFragmentViewModel.Inputs = this
```
Comment (Rcureton) [inline app/src/main/java/com/kickstarter/viewmodels/NewCardFragmentViewModel.kt]: You can just do `val inputs: Inputs = this`

File: app/src/main/java/com/kickstarter/viewmodels/NewCardFragmentViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        val outputs: NewCardFragmentViewModel.Outputs = this
+
+        private val apolloClient: ApolloClientType = this.environment.apolloClient()
```
Comment (Rcureton) [inline app/src/main/java/com/kickstarter/viewmodels/NewCardFragmentViewModel.kt]: You can use type inference with these variables

File: app/src/main/java/com/kickstarter/viewmodels/NewCardFragmentViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                    .compose(values())
+                    .subscribe { this.success.onNext(null) }
+
+            saveCardNotification
+                    .compose(Transformers.errors())
+                    .subscribe { this.error.onNext(it.localizedMessage) }
+
+        }
+
+        private fun storeNameAndPostalCode(cardForm: CardForm): Card {
```
Comment (Rcureton) [inline app/src/main/java/com/kickstarter/viewmodels/NewCardFragmentViewModel.kt]: Shouldn't these `private fun` be below the overrides?

File: app/src/main/java/com/kickstarter/viewmodels/NewCardFragmentViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            private fun isValidCard(card: Card?): Boolean {
+                return card != null && card.validateNumber() && card.validateExpiryDate() && card.validateCVC()
+            }
+
+            private fun isNotEmpty(s: String): Boolean {
+                return !s.isEmpty()
+            }
+        }
+    }
+}
```
Comment (Rcureton) [inline app/src/main/java/com/kickstarter/viewmodels/NewCardFragmentViewModel.kt]: NL

File: app/src/main/res/layout/fragment_new_card.xml [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+          </android.support.design.widget.TextInputLayout>
+
+        </LinearLayout>
+
+      </LinearLayout>
+
+      <FrameLayout
+        android:id="@+id/progress_bar"
+        android:layout_width="match_parent"
+        android:layout_height="4dp"
```
Comment (Rcureton) [inline app/src/main/res/layout/fragment_new_card.xml]: Maybe a nit but it's a loose dimen 🙃

File: app/src/main/java/com/kickstarter/services/KSApolloClient.kt [hunk 6]
```diff
@@ -35,6 +37,37 @@ class KSApolloClient(val service: ApolloClient) : ApolloClientType {
...
+                            ps.onError(exception)
+                        }
+
+                        override fun onResponse(response: Response<SavePaymentMethodMutation.Data>) {
+                            if (response.hasErrors()) {
+                                ps.onError(Exception(response.errors().first().message()))
+                            }
+                            //why wouldn't this just be an error?
+                            val createPaymentSource = response.data()?.createPaymentSource()
+                            if (!createPaymentSource?.isSuccessful!!) {
```
[the remark this answers is not in the document]
Comment (eoji) [reply app/src/main/java/com/kickstarter/services/KSApolloClient.kt]: yes 😕

File: app/src/main/java/com/kickstarter/viewmodels/NewCardFragmentViewModel.kt (hunk 1 above)
Comment (eoji) [reply app/src/main/java/com/kickstarter/viewmodels/NewCardFragmentViewModel.kt]: Nice catch! Missed it when I moved it out of the Activity

File: app/src/main/res/layout/fragment_new_card.xml (hunk 5 above)
Comment (eoji) [reply app/src/main/res/layout/fragment_new_card.xml]: Nits are still valid feedback!

**Pull request merged**